core/src/main/java/org/apache/calcite/rel/rules/MinusToDistinctRule.java

```diff
@@ -0,0 +1,184 @@
+        .withOperandFor(minusClass));
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Minus minus = call.rel(0);
+
+    if (minus.all) {
+      // Nothing we can do
+      return;
+    }
+
+    final RelOptCluster cluster = minus.getCluster();
+    final RelBuilder relBuilder = call.builder();
+    final RexBuilder rexBuilder = cluster.getRexBuilder();
+    final int branchCount = minus.getInputs().size();
+
+    // For each child branch in minus,add a column which indicates branch index
+    //
+    // e.g. select EMPNO from emp -> select EMPNO,0 from emp
+    // 0 indicates that it comes from minus the first child branch
+    for (int i = 0; i < branchCount; i++) {
+      relBuilder.push(minus.getInput(i));
+      relBuilder.projectPlus(relBuilder.literal(new BigDecimal(i)));
+    }
+
+    // create a union above all the branches
+    relBuilder.union(true, branchCount);
+
+    final RelNode union = relBuilder.peek();
+    final int originalFieldCnt = union.getRowType().getFieldCount() - 1;
+
+    ImmutableList.Builder<RexNode> projects = ImmutableList.builder();
+    // skip the branch index column
+    projects.addAll(Util.first(relBuilder.fields(), originalFieldCnt));
+
+    // On top of the Union, add a Project and add branch cnt number boolean columns
+    //
+    // e.g. LogicalProject(EMPNO=[$0], $f1=[=($1, 0)], $f2=[=($1, 1)], $f3=[=($1, 2)])
+    // $f1,$f2,$f3 are the boolean indicate whether it comes from the corresponding branch
+    for (int i = 0; i < branchCount; i++) {
+      projects.add(
+          relBuilder.equals(relBuilder.field(originalFieldCnt),
+              relBuilder.literal(new BigDecimal(i))));
+    }
+
+    relBuilder.project(projects.build());
+
+    // Add the count(*) filter $f1(..) for each branch
+    ImmutableList.Builder<RelBuilder.AggCall> aggCalls = ImmutableList.builder();
+    for (int i = 0; i < branchCount; i++) {
+      aggCalls.add(relBuilder.countStar(null).filter(relBuilder.field(originalFieldCnt + i)));
+    }
+
+    final ImmutableBitSet groupSet = ImmutableBitSet.range(originalFieldCnt);
+    relBuilder.aggregate(relBuilder.groupKey(groupSet), aggCalls.build());
+
+    ImmutableList.Builder<RexNode> filters = ImmutableList.builder();
+    for (int i = 0; i < branchCount; i++) {
+      SqlOperator operator =
+          i == 0
+              ? SqlStdOperatorTable.GREATER_THAN
+              : SqlStdOperatorTable.EQUALS;
+      filters
+          .add(
+              rexBuilder.makeCall(operator, relBuilder.field(originalFieldCnt + i),
+              relBuilder.literal(new BigDecimal(0))));
+    }
+
+    relBuilder.filter(filters.build());
+    relBuilder.project(Util.first(relBuilder.fields(), originalFieldCnt));
+    call.transformTo(relBuilder.build());
+  }
+
+  /**
+   * Rule configuration.
+   */
+  @Value.Immutable
+  public interface Config
+      extends RelRule.Config {
+    MinusToDistinctRule.Config DEFAULT = ImmutableMinusToDistinctRule.Config.of()
```

Comment:
No need for `MinusToDistinctRule.Config`, `Config` alone suffices and makes the code more readable, please fix also below.